perf/reader_test.go

```diff
@@ -85,43 +65,76 @@ func TestReaderSetDeadline(t *testing.T) {
 	}
 }
 
-func outputSamplesProg(sampleSizes ...int) (*ebpf.Program, *ebpf.Map, error) {
-	const bpfFCurrentCPU = 0xffffffff
+func outputSamples(tb testing.TB, events *ebpf.Map, sampleSizes ...byte) {
+	prog := outputSamplesProg(tb, events, sampleSizes...)
 
-	events, err := ebpf.NewMap(&ebpf.MapSpec{
-		Type: ebpf.PerfEventArray,
-	})
+	ret, _, err := prog.Test(internal.EmptyBPFContext)
+	testutils.SkipIfNotSupported(tb, err)
 	if err != nil {
-		return nil, nil, err
+		tb.Fatal(err)
+	}
+
+	if errno := syscall.Errno(-int32(ret)); errno != 0 {
+		tb.Fatal("Expected 0 as return value, got", errno)
 	}
+}
+
+// outputSamplesProg creates a program which submits a series of samples to a PerfEventArray.
+//
+// The format of each sample is:
+//
+//	index:   0    1    2    3    ... size - 1
+//	content: size id   0xff 0xff ... 0xff     [padding]
+//
+// padding is an implementation detail of the perf buffer and 1-7 bytes long. The
+// contents are undefined.
```

Comment:
Ah yes, that was based on a discussion with @ti-mo. Seems like padding is zero after all?

Reply:
Padding is not zeroed by the kernel and will contain garbage after it's wrapped the ring. I've revived this thread recently: https://lore.kernel.org/lkml/20200519132616.794171-1-timo@incline.eu as it was never merged.

See https://github.com/cilium/ebpf/pull/94#discussion_r424357585 for full context. I'd keep the test to make sure we're not making mistakes elsewhere.

Reply:
> Padding is not zeroed by the kernel and will contain garbage after it's wrapped the ring.

That's what's confusing me. We wrap the buffer in tests, and we check that the padding is zero. Yet the test doesn't fail. So either we're really lucky and don't get garbage by chance (how? maybe because our sample sizes don't have padding?) or it actually gives zero.

Reply:
Okay, the reason we don't see it in tests is that we never overwrite data in the regular perf buffer case. For the overwritable buffer I'm not so sure, maybe it's because the samples fit exactly?